data-plane/receiver-loom/src/main/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducer.java

```diff
@@ -15,6 +15,162 @@
+        this.producer = producer;
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.isClosed = new AtomicBoolean(false);
+        this.vertx = (VertxInternal) v;
+        this.ctx = vertx.getOrCreateContext();
+        ContextInternal ctxInt = ((ContextInternal) v.getOrCreateContext()).unwrap();
+        this.tracer = ProducerTracer.create(ctxInt.tracer());
+
+        sendFromQueueThread = Thread.ofVirtual().start(this::sendFromQueue);
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        Promise<RecordMetadata> promise = Promise.promise();
+        if (isClosed.get()) {
+            promise.fail("Producer is closed");
+        } else {
+            eventQueue.add(new RecordPromise<K, V>(record, promise));
+        }
+        return promise.future();
+    }
+
+    private void sendFromQueue() {
+        while (!isClosed.get() || !eventQueue.isEmpty()) {
+            try {
+                RecordPromise<K, V> recordPromise = eventQueue.take();
+                ProducerTracer.StartedSpan startedSpan =
+                        this.tracer == null ? null : this.tracer.prepareSendMessage(ctx, recordPromise.getRecord());
+
+                recordPromise
+                        .getPromise()
+                        .future()
+                        .onComplete(v -> {
+                            if (startedSpan != null) {
+                                startedSpan.finish(ctx);
+                            }
+                        })
+                        .onFailure(cause -> {
+                            if (startedSpan != null) {
+                                startedSpan.fail(ctx, cause);
+                            }
+                        });
+                try {
+                    producer.send(recordPromise.getRecord(), (metadata, exception) -> {
+                        if (exception != null) {
+                            recordPromise.getPromise().fail(exception);
+                        }
+                        recordPromise.getPromise().complete(metadata);
+                    });
+                } catch (final KafkaException exception) {
+                    recordPromise.getPromise().fail(exception);
+                }
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting for event queue to be populated.");
+                break;
+            }
+        }
+        logger.debug("Background thread finish.");
+    }
+
+    @Override
+    public Future<Void> close() {
+        Promise<Void> promise = Promise.promise();
+        this.isClosed.set(true);
+        Thread.ofVirtual().start(() -> {
+            try {
+                while (!eventQueue.isEmpty()) {
+                    Thread.sleep(2000L);
+                }
+                sendFromQueueThread.interrupt();
+                producer.close();
+                sendFromQueueThread.join();
```

Comment:
It seems more natural to interrupt and wait for it to terminate, what is the reason for the different order?

```suggestion
                sendFromQueueThread.join();
                producer.close();
```

Reply:
No, particular reason. okay I will change it.